[3.x] deps: support Laravel 12.x

calebdw opened the pull request:
Closes #2192

calebdw committed:
deps: allow Laravel 12.x

szepeviktor commented:
> Laravel 12 will be launched on February 24th, 2025.

calebdw committed:
test: add more descriptive error message

calebdw committed:
fix: test-case-extension integration test

calebdw committed:
fix: risky test not flushing error/exception handlers

calebdw commented:
@canvural, @szepeviktor, @ondrejmirtes,

This PR adds support for Laravel 12 as well as the `GenericStaticType` that's going to be added in the next PHPStan version (no need for #2191 anymore).

The addition of the new `GenericStaticType` means that the following extension are no longer needed (and they were in fact interfering with the `GenericStaticType` return types causing all the failures seen in #2191):
- src/ReturnTypes/EnumerableGenericStaticMethodDynamicStaticMethodReturnTypeExtension.php
- src/ReturnTypes/EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension.php

Since these extensions were removed I just went ahead and moved the data files from their respective test cases into the `GeneralTypeTest`.

ondrejmirtes commented:
Hi, is this in coordination with @canvural? Because I'm talking to him about GenericStaticType and it'd be great to release the fixes for it in both 2.x and 3.x, because the new type is part of both PHPStan 1.12.x and 2.1.x.

I'd like to release the new PHPStan versions as soon as possible, I certainly don't want to wait until Laravel 12 is released.

calebdw commented:
This PR can be released as soon as the new PHPStan version is released, it doesn't have to wait until Laravel 12 is released. It just adds support for it so folks can start preparing for it (like mentioned in #2192)

ondrejmirtes commented:
Alright! Can you please test the combo of this PR and dev PHPStan version with GenericStaticType on a real-world project, to see if it causes new issues or not?

ondrejmirtes commented:
So that I can still squeeze some bugfixes into PHPStan before release.

crynobone commented on tests/Type/GeneralTypeTest.php:
```diff
@@ -63,13 +72,20 @@ public static function dataFileAsserts(): iterable
         }
 
+        // 11.100 is an artificial constraint until 12.0.0 is released
+        if (version_compare(LARAVEL_VERSION, '11.0.0', '>=') && version_compare(LARAVEL_VERSION, '11.100.0', '<')) {
```
```suggestion
        if (laravel_version_compare('11.0.0', '>=') && laravel_version_compare('12.0.0', '<')) {
```

Don't forget to import `use function Orchestra\Testbench\laravel_version_compare;`

calebdw replied:
@crynobone, likely need to add 12.x here as well:


![image](https://github.com/user-attachments/assets/d9d5c5e9-e642-4710-9669-2edf1372dd9e)

crynobone commented on tests/Type/GeneralTypeTest.php:
```diff
@@ -63,13 +72,20 @@ public static function dataFileAsserts(): iterable
+            yield from self::gatherAssertTypes(__DIR__ . '/data/collection-generic-static-methods-l11.php');
+        }
+
+        if (version_compare(LARAVEL_VERSION, '12.0.0', '>=') || LARAVEL_VERSION === '12.x-dev') {
```
```suggestion
        if (laravel_version_compare('12.0.0', '>=')) {
```

calebdw committed:
chore: update alias to 3.x

calebdw commented:
> Can you please test the combo of this PR and dev PHPStan version with GenericStaticType on a real-world project, to see if it causes new issues or not?

@ondrejmirtes, I just executed phpstan in our ERP codebase and everything is still green!

crynobone commented on tests/Type/GeneralTypeTest.php:
```diff
@@ -63,13 +72,20 @@ public static function dataFileAsserts(): iterable
+            yield from self::gatherAssertTypes(__DIR__ . '/data/collection-generic-static-methods-l11.php');
+        }
+
+        if (laravel_version_compare('12.0.0', '>=') || LARAVEL_VERSION === '12.x-dev') {
```
```suggestion
        if (laravel_version_compare('12.0.0', '>=')) {
```

Testbench Core already check for `12.x-dev`: https://github.com/orchestral/testbench-core/blob/c3d080c533cff1960aa0468dd930cce76fdfea5e/src/functions.php#L417-L430

calebdw replied:
Ah sweet! must've been looking at the wrong version

canvural commented:
Hi,

Can you just keep the Laravel 12 support and drop the rest? Thank you.

calebdw commented:
Yes, I'll rebase this evening when I get the chance

calebdw commented:
@canvural, this has been rebased :+1: 


I went ahead and moved the data files from the deleted extension test cases to the `GeneralTypeTest` and removed the test cases. In the future we could presumably remove those data files as it's now testing PHPUnit and the framework and doesn't rely on any of our code---but it's good to leave in for now to prove that there's no breaking changes

canvural commented on composer.json:
```diff
@@ -17,25 +17,25 @@
-        "illuminate/http": "^11.15.0",
-        "illuminate/pipeline": "^11.15.0",
-        "illuminate/support": "^11.15.0",
+        "illuminate/console": "^11.41.3 || ^12.0",
+        "illuminate/container": "^11.41.3 || ^12.0",
+        "illuminate/contracts": "^11.41.3 || ^12.0",
+        "illuminate/database": "^11.41.3 || ^12.0",
+        "illuminate/http": "^11.41.3 || ^12.0",
+        "illuminate/pipeline": "^11.41.3 || ^12.0",
+        "illuminate/support": "^11.41.3 || ^12.0",
```
Why are we also increasing the minimum 11.x version?

calebdw replied:
The minimum 11.x version was updated to fix compatibility issues on higher 11.x versions.

In my opinion, it is much cleaner and maintainable to fix the issue and raise the minimum version than to create two **nearly identical** files and have an if condition in the GeneralTypeTest like you did here: https://github.com/larastan/larastan/commit/cef7cfdfe6cf0905823e932e24046175dc6d05bd

canvural replied:
I'd like to keep the support for wide range of versions.

calebdw replied:
I'll update since you insist, however, this is not very useful given that a `composer update` is going to update people to the latest 11.x---there's no reason for anyone to not be on the latest minor version.

canvural commented on composer.json:
```diff
@@ -17,25 +17,25 @@
         "doctrine/coding-standard": "^12.0",
         "nikic/php-parser": "^5.3",
-        "laravel/framework": "^11.15.0",
+        "laravel/framework": "^11.41.3 || ^12.0",
```
ditto

calebdw replied:
same as above

canvural commented on tests/Integration/IntegrationTest.php:
Why the changes in this file?

calebdw replied:
The line number changed because the constructor was removed

calebdw replied:
@canvural,

The other changes in this file are for developer quality of life when there's a test failure.

I got really tired of tests failing with the message: "Failed asserting that array has the key x"
- I don't know what the actual error is
- I don't know what test file it occurred in

These changes now print the exact file, error, and line number that occurred making it much easier to diagnose the problem.

canvural commented on tests/Integration/data/test-case-extension.php:
ditto

calebdw replied:
These changes were required because the PHPUnit TestCase (which this extends) has a final, strong-typed constructor:

![image](https://github.com/user-attachments/assets/f3669c7a-08e1-4f2f-ada5-4436499821c6)

canvural commented on tests/Rules/UnusedViewsRuleTest.php:
```diff
@@ -42,14 +42,6 @@ protected function getCollectors(): array
         ];
     }
 
-    protected function setUp(): void
-    {
-        parent::setUp();
-
-        // This is a workaround for a weird PHPStan container cache issue.
-        require __DIR__ . '/../../bootstrap.php';
-    }
-
```
Why is it removed?

calebdw replied:
This is no longer required and it was causing PHPUnit to report it as a risky test:

![image](https://github.com/user-attachments/assets/5d6a920c-a33e-42a1-b162-416c5b7bf098)

canvural replied:
hmm. But we do not need to use PHPUnit 11. Laravel 12 seems to still support PHPUnit 10

calebdw replied:
I allow both and there's no reason not to support it---it's going to have to be done eventually.

The fact also remains that this "weird PHPStan container cache issue" no longer seems to exist given that the tests are passing just fine

canvural commented on tests/Type/GeneralTypeTest.php:
```diff
@@ -6,7 +6,7 @@
 use PHPStan\Testing\TypeInferenceTestCase;
 
-use function version_compare;
+use function Orchestra\Testbench\laravel_version_compare;
```
What does this function do differently?

calebdw replied:
This function properly handles the `12.x-dev` version, otherwise you would have to do:

![image](https://github.com/user-attachments/assets/5fbc8f47-45a5-4606-b694-208f14f468f0)

canvural replied:
@crynobone Looks like this function is deprecated. Should we require sidekick here and use the function?

calebdw replied:
Done

calebdw replied:
Actually this is not doable given your constraint to stay on `11.15`:

![image](https://github.com/user-attachments/assets/10be1499-7ae2-4921-9770-7cf3efe3f823)

canvural reviewed (changes requested):
Hi,

There are too many changes that are not really related to the Laravel 12 compatibility itself. If I run locally with Laravel 12 I only get 3 errors and they are all related to `partition` method. I do not undertsand why would we need the other changes.

Also, looks like there are some conflicts. Let's fix those and keep the changes in the PR to minimal. It also has some unnecessary style changes. If PHP_CodeSniffer doesn't change we don't need to change it.

calebdw commented:
@canvural, this has been rebased and tests are passing.

> I do not undertsand why would we need the other changes.

I'm not making changes willy-nilly---there is a good reason for every change on this branch. If you're still confused on some changes then please ask.

> Also, looks like there are some conflicts. Let's fix those and keep the changes in the PR to minimal. It also has some unnecessary style changes. If PHP_CodeSniffer doesn't change we don't need to change it.

I have formatters in Neovim that autorun when I save. I usually try to avoid these changes but sometimes they slip through without me realizing it, although they shouldn't hurt anything

canvural commented:
> I have formatters in Neovim that autorun when I save. I usually try to avoid these changes but sometimes they slip through without me realizing it, although they shouldn't hurt anything

well they hurt the PR review process. I need to think if there was an actual change there or if it was only formatting. Like I said let's let the PHP_CodeSniffer to handle the formatting.

Anyway, thank you.

calebdw commented:
> well they hurt the PR review process. I need to think if there was an actual change there or if it was only formatting. Like I said let's let the PHP_CodeSniffer to handle the formatting.

I'll try to be better about not including changes like that in the future :+1: 

Thank you sir!

szepeviktor commented:
This PR is much nicer in its new, slim form.
